### ISSUE TITLE
Make `CPUOffloadOptimizer` easier to integrate into training frameworks

### ISSUE THREAD

**Issue by gau-nernst** [pytorch/ao#725]
Gotten quite a lot of discussion when I tried to add `CPUOffloadOptimizer` to torchtune https://github.com/pytorch/torchtune/pull/1351

The main problem is that to make it work with a config system in a training framework, for the `optimizer_class`, we need to parse a string class name into an actual optimizer class i.e. parse `"torch.optim.Adam"` into class `torch.optim.Adam`

None of the solutions in https://github.com/pytorch/torchtune/pull/1351 seems particularly satisfying. @felipemello1 suggested maybe we can update the code in ao instead. Although I don't think the optimizer in ao should do the string parsing, there are some ways I think we can do to improve the situation
- Add a default `optimizer_class=torch.optim.AdamW`
- Make subclasses of specific optimizers e.g.
```python
class AdamWCPUOffload(CPUOffloadOptimizer):
    def __init__(self, params, **kwargs):
        super.__init__(params, optimizer_class=torch.optim.AdamW, **kwargs)
```

Creating this issue for comments and discussion.

cc @msaroufim @SalmanMohammadi @ebsmothers

**Comment by felipemello1** [comment]
by the way, thanks for PR and all the discussion. I apologize that it hasnt landed yet :/

**Comment by gau-nernst** [comment]
> Doesnt seem scalable or flexible :/
> This is a could be stop gap, but we wouldnt be able to try new combinations, like AdamW8bit.

For CPU offload optimizer, having a fast CPU optimizer is crucial. Currently fused CPU Adam/AdamW in PyTorch core (>= 2.4) are the only "viable" optimizers that I know off (i.e. fast enough).

Other optimizers usually try to reduce memory usage e.g. Adafactor, AdamW8bit. Since we are already offloading optimizer state to CPU RAM, I don't think saving CPU RAM is particularly necessary? -> no big need to use other optimizers other than Adam/AdamW. Correct me if my impression is wrong. (other optimizers are also not "fast enough" on CPU)

**Comment by felipemello1** [comment]
Fair enough! AdamW8bit was bad example :P 

I am comfortable approving the PR as is. Salman had already approved. If Evan is ok, lets do it.

**Comment by gau-nernst** [comment]
To summarize, I think for now simply setting a default `optimizer_class=torch.optim.AdamW` should suffice for 2 reasons: (1) fast CPU implementation, and (2) battle-tested for convergence. Memory-efficient optimizers are not attractive in this case because optimizer states are already on CPU. When there are good alternatives, we can reconsider how to make this easier to extend for usage within a training library. (though personally I think parsing logic should stay in the training library 😄)

**Comment by felipemello1** [comment]
I am fine with this solution (having AdamW as default)!

Just a thought: It has been in the back of my mind to try a scheduler free optimizer: https://github.com/facebookresearch/schedule_free

The benefit is that the user is free to train for however long they want without worrying about tuning the lr decay / number of steps. The performance is usually equal or higher than highly tuned cosine annealing. But this is a problem for another time.

**Comment by gau-nernst** [comment]
For future reference. This is the fused CPU Adam implementation in a single file. Possibly straight-forward to adapt for other optimizers.

https://github.com/pytorch/pytorch/blob/519342962da2aa936e13d5af8cc5cda2084eda15/aten/src/ATen/native/cpu/FusedAdamKernel.cpp

**Comment by msaroufim** [comment]
We'd likely want to code generate all the other optimizers. So might be more fun to improve the torch compile cpu backend
